designs/statefulset-disruption.md

```diff
@@ -0,0 +1,891 @@
+
+For the purpose of this document, assume volumes have already been created and will never be deleted.
+
+<details open>
+<summary>If you want to dive one level deeper, open the dropdown to see the following diagram of what happens between pod eviction and volume detachment</summary>
+<img src="https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/fc577f2c-2de0-4aee-8daa-110a1eb0b990" alt="Stateful Pod Termination" class="inline"/>
+</details>
+
+### Ideal Disruption Flow for Stateful Workloads
+
+In order for a stateful pods to smoothly migrate from the terminating node to another node, the following steps must occur in order:
+
+0. Node marked for deletion
+1. Stateful pods must enter `terminated` state
+2. Volumes must be confirmed as unmounted (By CSI Node Service)
+3. Volumes must be confirmed as detached from instance (By AttachDetach & CSI Controllers)
+4. Karpenter terminates EC2 Instance
+5. Karpenter deletes finalizer on Node
+
+See the following diagram for a more detailed sequence of events.  
+
+![ideal](https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/ae5267e1-fe2d-455f-9199-60cee772f0ae)
+
+### Problems
+
+Today, with customers with default Karpenter `v0.37.0` and EBS CSI Driver `v1.31.0` configurations may experience two different kinds of delays once their disrupted stateful workloads are scheduled on a new node. 
+
+**Problem A. If step 2 doesn't happen, there will be a 6+ minute delay.**
+
+If volumes are not *confirmed* as unmounted *by CSI Node Service*, Kubernetes cannot confirm volumes are not in use and will wait 6 minutes and confirm node is unhealthy before treating the volume as unmounted. See [EBS CSI 6-minute delay FAQ](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/faq.md#6-minute-delays-in-attaching-volumes) for more context. [^6] 
+
+Customers will see the following event (Note the 6+ minute delay): 
+
+```
+Warning  FailedAttachVolume      6m51s              attachdetach-controller  Multi-Attach error for volume "pvc-123" Volume is already exclusively attached to one node and can't be attached to another
+```
+
+**Problem B. If step 3 doesn't happen before step 4, there will be a 1+ minute delay**
+
+If Karpenter calls EC2 TerminateInstance **before** EC2 DetachVolume calls from EBS CSI Driver Controller pod finish, then the volumes won't be detached **until the old instance terminates**.This delay depends on how long it takes the underlying instance to enter the `terminated` state, which depends on the instance type. Typically 1 minute for `m5a.large`, up to 15 minutes for certain Metals/GPU/Windows instances. See [appendix D1](#d1-ec2-termination--ec2-detachvolume-relationship-) for more context. 
```

Comment:
Mention "From my brief experiments, Typically 1 minute for 5ma.large..."

Reply:
This variability is really unfortunate. I wonder if you see more consistent behavior with `ec2:StopInstances`? I assume you can detach a volume when an instance is stopped? If so, maybe stopping prior to termination could at least put a more reasonable upper bound on the delay in this case

Reply:
Thanks for this @cartermckinnon, two notes for you. 

> I assume you can detach a volume when an instance is stopped? If so, maybe stopping prior to termination would at least put a more reasonable upper bound on the delay in this case

1. Great idea on `ec2:StopInstances`, but I'm not sure if the extra step is worth it considering that one can only detach a volume after instance reaches `stopped` state (which would take an extra ~8 seconds).  I measured the difference between stopping/terminating durations for a couple different instance types [here](https://github.com/aws/karpenter-provider-aws/blob/95c2945d8bdb8ea6b3ce6e0d3f1f787e78f7a192/designs/statefulset-disruption.md#instance-stoppedterminated-times). Looks like `stopped` is 10-40 seconds faster depending on instance type or EC2 quirks. 

2. My original document was misleading when talking about GPU/Windows instance termination times. They may be closer to linux termination times than metal instances. I have added an [instance termination timings section](https://github.com/aws/karpenter-provider-aws/blob/95c2945d8bdb8ea6b3ce6e0d3f1f787e78f7a192/designs/statefulset-disruption.md#instance-stoppedterminated-times) where I tested a few instance types. I was relying on hearsay timings when I should have dived deeper before the meeting, my apologies.  

> The guest OS can take a long time to complete shutting down.

I can try digging into this with the folks at EC2.